server/bucket_notifications/views.py

```diff
@@ -0,0 +1,153 @@
+import os
+
+from bson.objectid import ObjectId
+from girder import logger
+from girder.api import access
+from girder.api.describe import Description, autoDescribeRoute
+from girder.api.rest import Resource
+from girder.constants import SortDir
+from girder.exceptions import RestException
+from girder.models.assetstore import Assetstore
+from girder.models.folder import Folder
+from girder.models.user import User
+
+from dive_utils.types import AssetstoreModel, GirderModel
+
+from .constants import AssetstoreRuleMarker
+from .models import (
+    GCSNotificationRecord,
+    GCSPushNotificationPayload,
+    NotificationRouterRule,
+)
+
+
+class BucketNotification(Resource):
+    def __init__(self):
+        super(BucketNotification, self).__init__()
+        self.resourceName = "bucket_notifications"
+
+        self.route("POST", ("routing", ":id"), self.set_notification_routing)
+        self.route("GET", ("gcs",), self.gcs_get_records)
+        self.route("POST", ('gcs',), self.gcs_save_record)
+
+    @staticmethod
+    def processNotification(
+        store: AssetstoreModel, rootFolder: GirderModel, importPath: str
+    ):
+        """
+        Import at proper location
+        """
+        if rootFolder is None:
+            raise RestException('Root folder missing')
+        # Find the correct import location given rootFolder and importPath
+        owner = User().findOne(
+            {'_id': ObjectId(rootFolder['creatorId'] or rootFolder['baseParentId'])}
+        )
+        if owner is None:
+            raise RestException('Root has no owner')
+
+        target = rootFolder
+
+        importPath = importPath.lstrip('/')
+        realImportPathDir, _ = os.path.split(importPath)
+        storePrefix = str(store['prefix']).lstrip('/')
+        common_path = os.path.commonpath([storePrefix, importPath])
+
+        realImportPath = common_path
+        path_from_root = importPath[len(common_path) :].lstrip('/')
+        path_from_root_list, _ = os.path.split(path_from_root)
```

Comment:
This logic gets a little hairy, and I haven't commented it yet.  I'll try and do that soon.